src/simulation.py

```diff
@@ -0,0 +1,233 @@
+        Keywords:
+            - Base image
+            - Output image
+            - Max generations
+            - Max iterations
+            - Stagnation limit
+            - Number evaluations
+            - Height
+            - Width
+            - Number Verticies
+        """
+
+        self.base_image = kwargs.get("b_image")
+        self.output_image = kwargs.get("o_image")
+        self.max_generations: int = kwargs.get("m_gen", 10)
+        self.max_iterations: int = kwargs.get("m_iter", 10)
+        self.stagnation_limit: int = kwargs.get("stag_lim", 10)
+        self.n_verticies: int = kwargs.get("n_vert", 3)
+        self.num_evals: int = kwargs.get("n_evals", 50000)
+        # NOTE: this value is from the paper
+
+        # Derived class variables
+        self.height, self.width = self.base_image.shape
+        self.canvas = Canvas(list())
+        self.counter = 0
+
+        self.canvas = self.create_polygon(self.canvas)
+
+    def update_probabilities(
+        self,
+    ):
+        """
+        Choose a polygon to mutate weighted by the sequence probabilities
+        """
+        num = self.canvas.how_many()
+
+        # Generate Geometric series of probabilities for the sequence
+        probabilities = array([1 / (2 ** (1 + x)) for x in range(num)])
+
+        # Normalize the probabilities to 1
+        n_probabilities = probabilities / np.sum(probabilities)
+        self.probabilities = n_probabilities
+        return n_probabilities
+
+    def norm_opti_probs(
+        self,
+    ):
+        """
+        Create a uniform distribution with mg elements, where mg is the max
+        generations / max number of polygons
+        """
+        self.probabilities = [1 / self.max_generations] * self.max_generations
+        return self.probabilities
+
+    def eval_loss(self, image):
+        """
+        Evaluate an image to the base_image and return the SAD
+        """
+        return sad(self.base_image, image)
+
+    def cc_loss(self, parent, child):
+        """
+        Compute and compare loss
+        """
+
+        # compute the loss of the child iteration with the parent
+        l_parent = self.eval_loss(parent)
+        l_child = self.eval_loss(child)
+
+        # compare loss
+        if l_child < l_parent:
+            self.counter = 0
+        else:
+            self.counter += 1
+
+        return l_parent, l_child
+
+    def select(
+        self,
+    ):
+        """
+        Using probabilities, randomly select and return a polygon from the
+        canvas sequence and its index
+        """
+        indx, self.polygon_i = np.random.choice(
+            list(enumerate(self.canvas.sequence)), p=self.probabilities
+        )
+
+        return indx, self.polygon_i
+
+    def create_polygon(self, c: Canvas):
+        """
+        Create polygon and add it to the canvas
+        """
+
+        c = add_polygon(canvas=c, polygon=polygon_init(id=self.canvas.how_many()))
+
+        self.counter = 0
```

Comment:
This should only happen if it's not a re-initialization. We should confirm that we're not using this for reinit step

Reply:
I have remove the counter reset so the loop behaves as intended -- good catch!